Source/WebKit/Platform/IPC/Connection.cpp

```diff
@@ -102,13 +100,21 @@ class Connection::SyncMessageState {
     // Add matching pending messages to the provided MessageReceiveQueue.
     void enqueueMatchingMessages(Connection&, MessageReceiveQueue&, const ReceiverMatcher&);
 
-private:
-    friend class LazyNeverDestroyed<Connection::SyncMessageState>;
-    SyncMessageState() = default;
-
     // Dispatch pending sync messages for given connection.
     void dispatchMessagesAndResetDidScheduleDispatchMessagesForConnection(Connection&);
 
+private:
+    explicit SyncMessageState(RunLoop& runLoop)
+        : m_runLoop(runLoop)
+    {
+    }
+    static Lock syncMessageStateMapLock;
+    static HashMap<RunLoop*, SyncMessageState*>& syncMessageStateMap() WTF_REQUIRES_LOCK(syncMessageStateMapLock)
+    {
+        static NeverDestroyed<HashMap<RunLoop*, SyncMessageState*>> map;
```

Comment:
Not here, destruction is not available to unique_ptr because there is no unique owner for the instance. The intention is that nobody can call the destructor except the deleter, because in no other way the deletion can be correct.

It's a reference counted instance with non-copyable references. SyncMessageState::get() is the factory for the references. I'll make sure the ::get() will already return the unique_ptr, though

Each reference has an unique owner. The unique_ptr is used to hold each reference via custom unique_ptr.

Using the unique_ptr here is more of an artefact for the class being inside Connection.h . If the SyncMessageState would be in its own header, it could be called something like SyncMessageStateHolder as there would be more room to type up the ceremony in the individual header file.